### PR TITLE
MLmodels

### PR DESCRIPTION
**Pull request by YazanGhafir**
* Created MLmodels map, exported Lasso, ridge models as separate model classes
* Implemented ElasticNets, DecisionTrees, RandomForest models. 
* Enhanced the implementation of the models with automatic data loading, result printing and 
* integrated learning curve plotting in all of the models.
* Created Models Runner class that runs all the models or specific models as requested and print the results and plot the curves. 
* Created also Data loader class to centralize the reading of the data, printing it, and integrated implementation of PCA to plot the data points in 2D, to see if there are eventual correlations between the data points.  
* Implemented CompareModels function that compares the test error of the training of the models.

### CODE REVIEW

File: models/.gitignore [hunk 1]
```diff
@@ -2,3 +2,4 @@
 
 data/
```
Comment (vifraa) [inline models/.gitignore]:
```suggestion

```

Comment (vifraa) [reply models/.gitignore]: Fix på merge conflict.

File: models/MLmodels/ModelsRunner.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    data = DataSample()
+
+    models_dict = {
+        'Lasso': Lasso(data), 'Ridge': Ridge(data), 'ElasticNet': ElasticNet(data),
+        'DecisionTree': DecisionTree(data), 'RandomForest': RandomForest(data),
+        'NeuralNetwork': NeuralNetwork(data)
+    }
+
+    def __init__(self):
+        pass
```
Comment (vifraa) [inline models/MLmodels/ModelsRunner.py:23]: Skapa en parameter för att skicka in data som inte är DataSample, likt vad som är gjort på modellerna.

File: models/MLmodels/NeuralNetwork.py [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+            self.data = DataSample()
+        else:
+            self.data = data
+        self.regression()
+
+    def regression(self):
+        pass
+
+    def plot_learning_curves(self):
+        pass
```
Comment (vifraa) [inline models/MLmodels/NeuralNetwork.py]: Ej implementerade.

File: models/MLmodels/DataReader.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
+import pandas as pd
+from IPython.display import display
+from sklearn.model_selection import train_test_split
+from visualizers import model_plotter
+
+class DataSample:
```
Comment (vifraa) [inline models/MLmodels/DataReader.py:6]: Alla modeller använder datan på en viss struktur som implicerar att den måste vara i en klass likt denna. För att man ska kunna trycka in data på ett dynamiskt sätt skapa gärna en klass utöver den här där man som argument i konstruktorn kan skicka in X och Y, och att det enda som klassen sen gör är att shuffla, splitta och spara på objektet för att användningen ska fungera i modeller.

File: models/MLmodels/DecisionTree.py [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    def __init__(self, data=None):
+        if data is None:
+            self.data = DataSample()
+        else:
+            self.data = data
+
+    def regression(self):
+        DecisionTreeM = tree.DecisionTreeRegressor()
+        DecisionTreeM.fit(self.data.Xtrain, self.data.Ytrain)
+
```
Comment (vifraa) [inline models/MLmodels/DecisionTree.py]: Måste finnas sätt att komma åt den underliggande tränade modellen för att kunna sparas ner för användning i recommendation engine.

File: models/MLmodels/ElasticNets.py [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+    def __init__(self, data=None):
+        if data is None:
+            self.data = DataSample()
+        else:
+            self.data = data
+
+    def regression(self):
+        eNet = ElasticNetModel(alpha=1.0)
+        eNet.fit(self.data.Xtrain, self.data.Ytrain)
+
```
Comment (vifraa) [inline models/MLmodels/ElasticNets.py]: Måste finnas sätt att komma åt den underliggande tränade modellen för att kunna sparas ner för användning i recommendation engine.

File: models/MLmodels/Lasso.py [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+class Lasso:
+    def __init__(self, data=None):
+        if data is None:
+            self.data = DataSample()
+        else:
+            self.data = data
+
+    def regression(self):
+        lasso = linear_model.Lasso(alpha=0.1)
+        lasso.fit(self.data.Xtrain, self.data.Ytrain)
```
Comment (vifraa) [inline models/MLmodels/Lasso.py]: Måste finnas sätt att komma åt den underliggande tränade modellen för att kunna sparas ner för användning i recommendation engine.

File: models/MLmodels/RandomForest.py [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+    def __init__(self, data=None):
+        if data is None:
+            self.data = DataSample()
+        else:
+            self.data = data
+
+    def regression(self):
+        RandomForestM = RandomForestRegressor(max_depth=10, random_state=0)
+        RandomForestM.fit(self.data.Xtrain, self.data.Ytrain)
+
```
Comment (vifraa) [inline models/MLmodels/RandomForest.py]: Måste finnas sätt att komma åt den underliggande tränade modellen för att kunna sparas ner för användning i recommendation engine.

File: models/MLmodels/Ridge.py [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+    def __init__(self, data=None):
+        if data is None:
+            self.data = DataSample()
+        else:
+            self.data = data
+
+    def regression(self):
+        ridge = RidgeModel(alpha=1.0)
+        ridge.fit(self.data.Xtrain, self.data.Ytrain)
+
```
Comment (vifraa) [inline models/MLmodels/Ridge.py]: Måste finnas sätt att komma åt den underliggande tränade modellen för att kunna sparas ner för användning i recommendation engine.

**Review by vifraa — CHANGES_REQUESTED** [review]
Överlag väldigt bra! Snyggt jobbat!!!

Drog lite kommentarer på vad som behöver ändras för att det ska kunna användas där viktigaste två är:
1. Klass för data där man kan skicka in X och Y. (Eftersom modellerna kräver ett objekt med datan liggandes på speciell struktur.)
2. Sätt att komma åt den tränade modellen från klassen. Utan detta går det inte att spara ner och användas i recommendation engine.

**Comment by YazanGhafir** [comment]
All the commens has been resolved and more is added such:

 An object of any class of the models can be instantiated in one of the following ways:

       * using path, ex: Ridge(path=GIVEN_PATH) then the constructor will read
            the data (csv) in the given path however its sized, just that the target is
            the located as the last column, then it partition, shuffle and instantiate the date to:
            self.data.Xtrain, self.data.Ytrain, self.data.Xtest, self.data.Ytest

       * using X and Y ex: Ridge(X=GIVEN_X, Y=GIVEN_Y) then the constructor will
            create data model with the given X and Y and partition it also to:
            self.data.Xtrain, self.data.Ytrain, self.data.Xtest, self.data.Ytest

       * using already partitioned data then the constructor will
            create data model with the given data and partition it also to:
            self.data.Xtrain, self.data.Ytrain, self.data.Xtest, self.data.Ytest

       * if nothing of the above passed as argument ex. Ridge() then
            the constructor will read ready data using the Data_sample class
            in DataReader.py and have it in the same format specified above
